src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -724,7 +724,20 @@ void BaseCouplingScheme::advanceTXTWriters()
 
 bool BaseCouplingScheme::reachedEndOfTimeWindow()
 {
-  return math::equals(getNextTimeStepMaxSize(), 0.0, _eps) || not hasTimeWindowSize();
+  if (math::equals(getNextTimeStepMaxSize(), 0.0, _minTimeStepSize) && getNextTimeStepMaxSize() > math::NUMERICAL_ZERO_DIFFERENCE && hasTimeWindowSize()) {
+    for (auto &data : _allData | boost::adaptors::map_values) {
+      auto   timeStepStorage = data->timeStepsStorage();
+      auto   values          = timeStepStorage.getTimesAndValues();
+      auto   ts              = values.first;
+      double endOfTimeWindow = getTime() - _computedTimeWindowPart + _timeWindowSize;
+      if (ts(ts.size() - 1) < endOfTimeWindow - math::NUMERICAL_ZERO_DIFFERENCE) {
+        data->setSampleAtTime(getTime() - _computedTimeWindowPart + _timeWindowSize, timeStepStorage.getSampleAtOrAfter(ts(ts.size() - 1)));
```

Comment:
Is this the sample you are duplicating? I think a comment would be helpful. Why do you duplicate the sample? Why not shift the timestamp of the sample at `ts(ts.size() - 1)` to `endOfTimeWindow` instead? (This is just an alternative, but if I think about higher order interpolants maybe a valid one? No need to implement this alternative, but a comment would help)

Reply:
I did not find a way to return a non const reference to the last sample, which means that I should not be able to modify the last sample directly if I have understood the cpp syntax correctly. Otherwise I think that moving the sample would have been a more elegant solution.

Reply:
Correct. There is currently no possibility to update samples or their timestamp. Like I said: Just an alternative idea and if there is no obvious benefit, let's go for the "duplication solution" as this is possible with what we already have.